src/DotNet.Status.Web/Controllers/AzurePipelinesController.cs

```diff
@@ -0,0 +1,375 @@
+                    b.Append("](");
+                    b.Append(url);
+                    b.Append(") - ");
+                }
+                else
+                {
+                    b.Append("- ");
+                    b.Append(change.Id, 0, Math.Min(8, change.Id.Length));
+                    b.Append(" - ");
+                }
+
+                b.Append(change.Author.DisplayName);
+                b.Append(" - ");
+                b.AppendLine(change.Message);
+            }
+
+            if (more > 0)
+            {
+                b.Append("- ... and ");
+                b.Append(more);
+                b.AppendLine(" more ...");
+            }
+
+            return b.ToString();
+        }
+
+        private async Task<string> BuildTimelineMessage(Build build)
+        {
+            Timeline timeline = await Client.GetTimelineAsync(build.Project.Name, build.Id, CancellationToken.None);
+            StringBuilder builder = new StringBuilder();
+            IEnumerable<TimelineRecord> stages = timeline?.Records.OrEmpty().Where(r => r.Type == "Stage");
+            foreach (TimelineRecord stage in stages.OrEmpty())
+            {
+                var issues = new List<(TimelineIssue issue, string log)>();
+                PopulateIssuesUnder(stage, timeline, issues);
+
+                if (issues.Count == 0)
+                {
+                    continue;
+                }
+
+                builder.Append("#### ");
+                builder.AppendLine(stage.Name);
+
+                foreach ((TimelineIssue issue, string log) in issues)
+                {
+                    switch (issue.Type)
+                    {
+                        case "error":
+                            builder.Append("- :x: - ");
+                            break;
+                        case "warning":
+                            builder.Append("- :warning: - ");
+                            break;
+                    }
+
+                    if (Uri.IsWellFormedUriString(log, UriKind.Absolute))
+                    {
+                        builder.Append("[[Log]](");
+                        builder.Append(log);
+                        builder.Append(") - ");
+                    }
+
+                    builder.AppendLine(issue.Message);
+                    builder.AppendLine();
+                }
+            }
+
+            return builder.ToString();
+        }
+
+        private void PopulateIssuesUnder(TimelineRecord parent, Timeline timeline, List<(TimelineIssue issue, string log)> issues)
+        {
+            foreach (var child in timeline.Records.Where(r => r.ParentId == parent.Id).OrderBy(r => r.Order))
+            {
+                if (child.Issues != null)
+                {
+                    issues.AddRange(child.Issues.Select(i => (i, child.Log.Url)));
+                }
+
+                PopulateIssuesUnder(child, timeline, issues);
+            }
+        }
+
+
+        /// <summary>
```

Comment:
Perhaps put these classes in a separate file.

Reply:
They are nested classes (so that they don't leak out into the parent namespace).  I could make this class partial and then have a bunch of weird partial files that each contain a single nested class, but that looks sort of weird...